### ISSUE TITLE
[BUG] Slow down of Pytorch Forecasting NNs on new arch

### ISSUE THREAD

**Issue by alex-hse-repository** [tinkoff-ai/etna#999]
### 🐛 Bug Report

Pytorch Forecasting NNs on new arch works 10x slower than in the old version. 

### Expected behavior

1. Pytorch Forecasting NNs on new arch works the same time as in the old version. 
2. To check this run script in etna-experiments benchmark_tsdataset/issue-992 for TFT and DeepAR

### Code(old arch)
```
import torch
import random
import pandas as pd
import numpy as np
from etna.datasets.tsdataset import TSDataset
from etna.pipeline import Pipeline
from etna.metrics import SMAPE, MAPE, MAE
from etna.transforms import DateFlagsTransform
from etna.transforms import PytorchForecastingTransform
from etna.models.nn import TFTModel

HORIZON = 7


def set_seed(seed: int = 42):
    """Set random seed for reproducibility."""
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)


set_seed()

original_df = pd.read_csv("examples/data/example_dataset.csv")
df = TSDataset.to_dataset(original_df)
ts = TSDataset(df, freq="D")

dft = DateFlagsTransform(day_number_in_week=True, day_number_in_month=False, out_column="regressor_dateflag")
pft = PytorchForecastingTransform(
    max_encoder_length=21,
    min_encoder_length=21,
    max_prediction_length=HORIZON,
    time_varying_known_reals=["time_idx"],
    time_varying_known_categoricals=["regressor_dateflag_day_number_in_week"],
    time_varying_unknown_reals=["target"],
    static_categoricals=["segment"],
    target_normalizer=None,
)


pipeline_tft = Pipeline(model=TFTModel(trainer_kwargs=dict(max_epochs=1)), transforms=[dft, pft], horizon=HORIZON)

metrics_tft, forecast_tft, fold_info_tft = pipeline_tft.backtest(
    ts, metrics=[SMAPE(), MAPE(), MAE()], n_folds=3, n_jobs=1
)
```
### Time(old_arch):
`[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:    7.4s finished`

### Code(new arch)
```
import torch
import random
import pandas as pd
import numpy as np
from etna.datasets.tsdataset import TSDataset
from etna.pipeline import Pipeline
from etna.metrics import SMAPE, MAPE, MAE
from etna.transforms import DateFlagsTransform
from etna.models.nn.utils import PytorchForecastingDatasetBuilder
from etna.models.nn import TFTModel

HORIZON = 7


def set_seed(seed: int = 42):
    """Set random seed for reproducibility."""
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)


set_seed()

original_df = pd.read_csv("examples/data/example_dataset.csv")
df = TSDataset.to_dataset(original_df)
ts = TSDataset(df, freq="D")


set_seed()

dft = DateFlagsTransform(day_number_in_week=True, day_number_in_month=False, out_column="regressor_dateflag")
model_tft = TFTModel(
    dataset_builder=PytorchForecastingDatasetBuilder(
        max_encoder_length=21,
        min_encoder_length=21,
        max_prediction_length=HORIZON,
        time_varying_known_reals=["time_idx"],
        time_varying_known_categoricals=["regressor_dateflag_day_number_in_week"],
        time_varying_unknown_reals=["target"],
        static_categoricals=["segment"],
        target_normalizer=None,
    ),
    trainer_params=dict(max_epochs=1),
)

pipeline_tft = Pipeline(model=model_tft, transforms=[dft], horizon=HORIZON)

metrics_tft, forecast_tft, fold_info_tft = pipeline_tft.backtest(
    ts, metrics=[SMAPE(), MAPE(), MAE()], n_folds=3, n_jobs=1
)
```
### Time(new_arch):
`[Parallel(n_jobs=1)]: Done   3 out of   3 | elapsed:  1.9min finished`

### Environment

"old_arch" == etna 1.13.0
"new_arch" == etna tsdataset_2 (commit https://github.com/tinkoff-ai/etna/commit/45cc3b5a8330e4960fa2ace2f979a73d06b09bdb)

### Checklist

- [X] Bug appears at the latest library version

**Comment by Mr-Geekman** [comment]
We have found a potential bottlenecks: [1](https://github.com/tinkoff-ai/etna/blob/master/etna/models/nn/utils.py#L137), [2](https://github.com/tinkoff-ai/etna/blob/master/etna/models/nn/utils.py#L195).

It should be confirmed that the problem is in that place and should be fixed.

**Comment by Mr-Geekman** [comment]
Closed by #1322
